### PR TITLE
[Quant][FX] lower functional conv_transpose ops

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97936
* #97311
* __->__ #97126

**Summary**
Support quantizing and lowering functional `conv_transpose1d`, `conv_transpose2d` and `conv_transpose3d`.
Please note that
- `conv_tranpose + relu` fusion is not supported. Remember to keep `relu` node in graph when lowering.
- `conv_tranpose` requires `per-tensor` scheme for weight. Use default `qconfig_mappings` instead of deprecated `qconfig_dict` for test cases.

**Test plan**
python test/test_quantization.py -k test_conv_transpose_not_reference
python test/test_quantization.py -k test_conv_transpose_reference
python test/test_quantization.py -k test_conv_transpose_relu_not_reference
python test/test_quantization.py -k test_conv_transpose_relu_reference

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 1]
```diff
@@ -790,6 +803,20 @@ def _lower_static_weighted_ref_functional(
...
+            if func_node.target == F.conv_transpose1d:
+                # Note prepack_args[5] is groups. See comments below.
+                for i in [2, 3, 4, 6]:
+                    if len(prepack_args) > i and isinstance(prepack_args[i], int):
+                        prepack_args[i] = (prepack_args[i],)
+            # swap dilation and groups
+            # prepack op has arguments: {w, b, stride, padding, output_padding, dilation, groups}
+            # transposed conv op has arguments: {x, w, b, stride, padding, output_padding, groups, dilation}
+            if (len(prepack_args) > 6):
+                prepack_args[5], prepack_args[6] = prepack_args[6], prepack_args[5]
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:819]: hopefully we can remove this

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 2]
```diff
@@ -293,11 +294,14 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
 # Mapping from a functional to lower to a 2-tuple of
 #   1) The quantized version of the op
 #   2) The quantized version of the op fused with relu, if it exists, else None
-STATIC_LOWER_FUNCTIONAL_MAP: Dict[Callable, Tuple[Callable, Callable]] = {
+STATIC_LOWER_FUNCTIONAL_MAP: Dict[Callable, Tuple[Callable, Union[Callable, None]]] = {
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py]: Thanks. It's fixed.

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 3]
```diff
@@ -790,6 +803,20 @@ def _lower_static_weighted_ref_functional(
...
+            # For conv_transpose1d, the stride, padding, and dilation args may be ints,
+            # in which case we need to convert them to tuples
+            if func_node.target == F.conv_transpose1d:
+                # Note prepack_args[5] is groups. See comments below.
+                for i in [2, 3, 4, 6]:
+                    if len(prepack_args) > i and isinstance(prepack_args[i], int):
+                        prepack_args[i] = (prepack_args[i],)
+            # swap dilation and groups
+            # prepack op has arguments: {w, b, stride, padding, output_padding, dilation, groups}
+            # transposed conv op has arguments: {x, w, b, stride, padding, output_padding, groups, dilation}
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py]: Thanks. It's fixed. I changed the order of args of prepack op.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 1 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py]: Thanks. They are removed after the changes above.

File: torch/ao/quantization/fx/utils.py [hunk 4]
```diff
@@ -152,6 +153,16 @@ def get_qconv_prepack_op(conv_op: Callable) -> Callable:
     assert prepack_op, "Didn't find prepack op for {}".format(conv_op)
     return prepack_op
 
+def get_qconv_transpose_prepack_op(conv_transpose_op: Callable) -> Callable:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:156]: nit: it feels like we can merge this helper function with get_qconv_prepack_op and just have a single helper function?

**Review by jerryzh168 — APPROVED** [review]
LGTM, had one more minor comment

**Comment by Xia-Weiwen** [comment]
> > LGTM, had one more minor comment
> 
> Thanks @jerryzh168. A problem is that backwards compatibility test failed due to the change of prepack op. How can I land this PR? Do I need some extra approval?

Maybe define new prepack ops with suffix `_2` to ensure backward compatibility?

**Review by jerryzh168 — COMMENTED** [review]
oh I see, maybe you can split the bc-breaking changes to a separate PR, you can update the bc compatiblity tests as well by setting `generate` to True for the failed test: https://github.com/pytorch/pytorch/blob/master/test/quantization/bc/test_backward_compatibility.py#L244 and after that I can help import the diff to internal and see if it breaks any internal tests as well

**Comment by Xia-Weiwen** [comment]
> oh I see, maybe you can split the bc-breaking changes to a separate PR, you can update the bc compatiblity tests as well by setting `generate` to True for the failed test: https://github.com/pytorch/pytorch/blob/master/test/quantization/bc/test_backward_compatibility.py#L244 and after that I can help import the diff to internal and see if it breaks any internal tests as well

OK. I will open a separate PR then. So, these lines will be added back in this PR:
```
            # swap dilation and groups
            # prepack op has arguments: {w, b, stride, padding, output_padding, dilation, groups}
            # transposed conv op has arguments: {x, w, b, stride, padding, output_padding, groups, dilation}
            if (len(prepack_args) > 6):
                prepack_args[5], prepack_args[6] = prepack_args[6], prepack_args[5]
```

**Comment by jerryzh168** [comment]
> > oh I see, maybe you can split the bc-breaking changes to a separate PR, you can update the bc compatiblity tests as well by setting `generate` to True for the failed test: https://github.com/pytorch/pytorch/blob/master/test/quantization/bc/test_backward_compatibility.py#L244 and after that I can help import the diff to internal and see if it breaks any internal tests as well
> 
> OK. I will open a separate PR then. So, these lines will be added back in this PR:
> 
> ```python
>             # swap dilation and groups
>             # prepack op has arguments: {w, b, stride, padding, output_padding, dilation, groups}
>             # transposed conv op has arguments: {x, w, b, stride, padding, output_padding, groups, dilation}
>             if (len(prepack_args) > 6):
>                 prepack_args[5], prepack_args[6] = prepack_args[6], prepack_args[5]
> ```

sounds good, we can unblock this PR by adding these back I think

**Pull request closed without merging**